### ISSUE TITLE
Require partial method declaration and definition signatures to match

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#45519]
Partial method declaration and definition signatures should match, other than in cases of backwards compatibility (based on email discussion with C# language design team).

In the following, `F1()`, `F2()`, `F3()` are existing scenarios in C#8, and `F4()`, `F5()`, `F6()` are new scenarios. Errors should be reported for the differences in `F4()`, `F5()`, `F6()`.
```C#
using System.Collections.Generic;

#nullable enable

partial class C
{
    partial void F1(dynamic o);
    partial void F1(object o) { } // C#8: no warning

    partial void F2(string? s);
    partial void F2(string s) { } // C#8: warning CS8611

    partial void F3((int x, int y) t);
    partial void F3((int, int) t) { } // C#8: error CS8142

    partial void F4(System.IntPtr i);
    partial void F4(nint i) { } // new: error

    public partial dynamic F5();
    public partial object F5() => null!; // new: error

    public partial void F6(IEnumerable<string?> s);
    public partial void F6(IEnumerable<string> s) { } // new: error
}
```

**Comment by cston** [comment]
cc @RikkiGibson

**Comment by RikkiGibson** [comment]
Related to #45301

**Comment by cston** [comment]
From LDM discussion:
- Report error CS8142 for tuple element name differences (unchanged)
- Report warning CS8611 for unsafe nullable differences (unchanged)
- Allow `#nullable enable` / `#nullable disable` differences without warning (unchanged)
- Report warning wave warning for all other differences (new)

**Comment by cston** [comment]
Re-opening and moving to 16.9.

**Comment by cston** [comment]
Fixed in https://github.com/dotnet/roslyn/pull/53352.
